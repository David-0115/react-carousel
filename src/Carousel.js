import { useState, useEffect } from "react";
import "./Carousel.css";
import Card from "./Card";


/** Carousel: displays images and arrows to navigate through them
 * 
 * Props:
 * - photos: array of {src, caption} objects
 * - title: string describing the collection of images
 * 
 * State:
 * - currCardIdx: integer for current card index
 * 
 * App --> Carousel --> Card
 */
function Carousel({ photos, title }) {
  const [currCardIdx, setCurrCardIdx] = useState(0);
  const [leftArrow, setLeftArrow] = useState(false);
  const [rightArrow, setRightArrow] = useState(true);

  const currCard = photos[currCardIdx];
  const total = photos.length;


  //Increments currCardIdx state by 1
  function goForward() {
    setCurrCardIdx(currCardIdx + 1);
  }

  //Decrements currCardIdx state by 1
  function goBack() {
    setCurrCardIdx(currCardIdx - 1)
  }
  useEffect(() => {
    const hideLeft = () => setLeftArrow(false);
    const showLeft = () => setLeftArrow(true);
    const hideRight = () => setRightArrow(false);
    const showRight = () => setRightArrow(true);

    currCardIdx > 0 ? showLeft() : hideLeft();
    currCardIdx === photos.length - 1 ? hideRight() : showRight();
  }, [currCardIdx, photos.length])


  return (
    <div className="Carousel">
      <h1>{title}</h1>
      <div className="Carousel-main">
        {leftArrow ? <i className="bi bi-arrow-left-circle" onClick={goBack} data-testid="left" /> : null}
        <Card
          caption={currCard.caption}
          src={currCard.src}
          currNum={currCardIdx + 1}
          totalNum={total}
        />
        {rightArrow ? <i className="bi bi-arrow-right-circle" onClick={goForward} data-testid="right" /> : null}

      </div>
    </div>
  );
}

export default Carousel;
