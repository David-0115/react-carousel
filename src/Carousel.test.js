import { render, fireEvent } from "@testing-library/react";
import Carousel from "./Carousel";
import TEST_IMAGES from "./_testCommon.js";
import Card from "./Card.js";

//Card smoke test
it("Card should render", () => {
  render(<Card caption="test" src={TEST_IMAGES[0]} currNum={1} totalNum={1} />)
});

//Card snapshot test
it("Card matches snapshot", () => {
  const { asFragment } = render(<Card caption="test" src={TEST_IMAGES[0]} currNum={1} totalNum={1} />);
  expect(asFragment()).toMatchSnapshot();
});

//Carousel smoke test
it("Carousel should render", () => {
  render(<Carousel photos={TEST_IMAGES} title="images for testing" />)
});

//Carousel snapshot test
it("Carousel should render", () => {
  const { asFragment } = render(<Carousel photos={TEST_IMAGES} title="images for testing" />);
  expect(asFragment()).toMatchSnapshot();
});

it("works when you click on the right arrow", function () {
  const { container } = render(
    <Carousel
      photos={TEST_IMAGES}
      title="images for testing"
    />
  );
  // expect the first image to show, but not the second
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).toBeInTheDocument();
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).not.toBeInTheDocument();

  // move forward in the carousel
  const rightArrow = container.querySelector(".bi-arrow-right-circle");
  fireEvent.click(rightArrow);

  // expect the second image to show, but not the first
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).not.toBeInTheDocument();
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).toBeInTheDocument();
});

it("works when you click on the left arrow", function () {
  const { container } = render(
    <Carousel
      photos={TEST_IMAGES}
      title="images for testing"
    />
  );

  // move forward in the carousel
  const rightArrow = container.querySelector(".bi-arrow-right-circle");
  fireEvent.click(rightArrow);

  // expect the second image to show, but not the first
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).not.toBeInTheDocument();
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).toBeInTheDocument();

  const leftArrow = container.querySelector(".bi-arrow-left-circle");
  fireEvent.click(leftArrow);
  //expect first image to show, but not the second. 
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).toBeInTheDocument();
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).not.toBeInTheDocument();
});

it("should not show left arrow at the start", () => {
  const { queryByTestId } = render(
    <Carousel
      photos={TEST_IMAGES}
      title="images for testing"
    />
  );

  expect(queryByTestId("left")).toBe(null);

});

it("should not show right arrow at the end", () => {
  const { container, queryByTestId } = render(
    <Carousel
      photos={TEST_IMAGES}
      title="images for testing"
    />
  );

  // move forward in the carousel
  const rightArrow = container.querySelector(".bi-arrow-right-circle");
  fireEvent.click(rightArrow);  //move to img2
  fireEvent.click(rightArrow);  //move to img3

  expect(queryByTestId("right")).toBe(null);

})
