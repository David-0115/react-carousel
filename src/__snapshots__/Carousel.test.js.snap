// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Card"
  >
    <h4
      class="Card-title"
    >
      test
    </h4>
    <img
      alt="test"
      class="Card-image"
      src="[object Object]"
    />
    <small
      class="Card-small"
    >
      Image 1 of 1.
    </small>
  </div>
</DocumentFragment>
`;

exports[`Carousel should render 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <h1>
      images for testing
    </h1>
    <div
      class="Carousel-main"
    >
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          testing image 1
        </h4>
        <img
          alt="testing image 1"
          class="Card-image"
          src="test1.com"
        />
        <small
          class="Card-small"
        >
          Image 1 of 3.
        </small>
      </div>
      <i
        class="bi bi-arrow-right-circle"
        data-testid="right"
      />
    </div>
  </div>
</DocumentFragment>
`;
